File: web/src/utils/visibility.ts
/**
 * 可见性标签工具函数
 * 根据 visibility、allowed_group_names 和 group_name 生成标签数组
 */

interface VisibilityItem {
  visibility: "public" | "friends" | "group";
  allowed_group_names?: string[];
  group_name?: string | null;
}

/**
 * 生成可见性标签数组
 * - public → ["公开"]；friends → ["好友"]（二者互斥，最多一个）
 * - allowed_group_names → 白名单群名，可与公开/好友叠加显示（"公开+群""好友+群"共存）
 * - group 可见且无白名单群名 → 回退 group_name（旧数据兼容）
 */
export function getVisibilityLabels(item: VisibilityItem): string[] {
  const labels: string[] = [];

  // 基础可见性标签（公开 / 好友 互斥）
  if (item.visibility === "public") {
    labels.push("公开");
  } else if (item.visibility === "friends") {
    labels.push("好友");
  }

  // 群可见：白名单群名（可与公开/好友叠加）
  if (item.allowed_group_names && item.allowed_group_names.length > 0) {
    labels.push(...item.allowed_group_names);
  } else if (item.visibility === "group" && item.group_name) {
    // 兼容旧数据：visibility=group 但无白名单群名时回退归属群名
    labels.push(item.group_name);
  }

  // 没有任何标签（异常数据），按 visibility 兜底
  if (labels.length === 0) {
    if (item.visibility === "friends") return ["好友"];
    if (item.visibility === "public") return ["公开"];
    return ["群可见"];
  }

  return labels;
}

File: flutter/lib/core/models/visibility.dart
/// 可见性（来源标识）共用模型 —— 对齐 web `Ayla/web/src/utils/visibility.ts`
/// 与 `api/types.ts` 的可见性字面量联合。
///
/// **为什么是共用件**：web 的 `visibility: "public" | "friends" | "group"` 与
/// `getVisibilityLabels()` 是**跨域共用**的——帖子（`Post.visibility`）、
/// 语音频道（`types.ts:872` `VoiceChannelDescriptor.visibility`）、直播频道
/// （`LiveChannelDescriptor.visibility`）用的是同一个联合与同一个生成函数
/// （`components/cards/cardData.ts` 的 `cardVisibilityLabels` 只是转发）。
/// 故这里承载共用语义，`post.dart` / 后续 voice/live 模型都从这里取。
///
/// 名称沿用历史（`AylaPostVisibility` / 原文件 `post.dart`），以免动既有调用点
/// ——与 `media_kind.dart` 的抽出同口径（2026-09-20 先例）。
library;

/// 可见性（`types.ts:1227`：public / friends / group）。
///
/// 注意：**group 是独立维度**（白名单 allowed_groups），与 public/friends 可叠加；
/// 后端单值字段只是准入映射，标签由 [aylaVisibilityLabels] 生成。
enum AylaPostVisibility {
  public,
  friends,
  group;

  static AylaPostVisibility? parse(String? raw) => switch (raw) {
        'public' => AylaPostVisibility.public,
        'friends' => AylaPostVisibility.friends,
        'group' => AylaPostVisibility.group,
        _ => null,
      };

  /// 后端字段值。
  String get wire => name;
}

/// 可见性标签（web `utils/visibility.ts getVisibilityLabels` 逐条同源）。
///
/// - public → ["公开"]；friends → ["好友"]（二者互斥，最多一个）；
/// - `allowedGroupNames` → 白名单群名（**可与公开/好友叠加**：「公开+群」）；
/// - group 可见且无白名单名 → 回退 `groupName`（旧数据兼容）；
/// - 一条都生不出来时按 visibility 兜底（friends→好友 / public→公开 / 否则群可见）。
List<String> aylaVisibilityLabels({
  AylaPostVisibility? visibility,
  List<String> allowedGroupNames = const <String>[],
  String? groupName,
}) {
  final List<String> labels = <String>[];
  if (visibility == AylaPostVisibility.public) {
    labels.add('公开');
  } else if (visibility == AylaPostVisibility.friends) {
    labels.add('好友');
  }
  if (allowedGroupNames.isNotEmpty) {
    labels.addAll(allowedGroupNames);
  } else if (visibility == AylaPostVisibility.group &&
      (groupName ?? '').isNotEmpty) {
    labels.add(groupName!);
  }
  if (labels.isEmpty) {
    if (visibility == AylaPostVisibility.friends) return <String>['好友'];
    if (visibility == AylaPostVisibility.public) return <String>['公开'];
    return <String>['群可见'];
  }
  return labels;
}
